lib/dubhash.ts

```diff
@@ -0,0 +1,194 @@
+// This module implements the "double hash" signature specification described in docs/signature_specification.md
+import * as crypto from 'crypto';
+import { SignatureResult } from './types';
+import { isBinary } from './utils/binary';
+
+const hashAlgorithm = 'md5';
+
+const regex = /=/g;
+
+enum LineEndingFormat {
+  LINUX,
+  WINDOWS,
+  UNKNOWN,
+}
+
+export const getDubHashSignature = (
+  filePath: string,
+  fileContents: Buffer,
+  getAltHash = false,
+): SignatureResult | null => {
+  // Unless getAltHash is true, we only need a single hash, AKA the "half double hash".
+  if (!getAltHash) {
+    return {
+      path: filePath,
+      hashes_ffm: [
+        {
+          data: getSingleDigest(fileContents),
+          format: 1,
+        },
+      ],
+    };
+  }
+
+  const [digest1, digest2] = getDubHashDigests(
+    fileContents,
+    isBinary(fileContents),
+  );
+  if (!digest1) throw Error(`Failed to generate hash of ${filePath}`);
+  if (digest1 && !digest2) {
+    return {
+      path: filePath,
+      hashes_ffm: [
+        {
+          data: digest1,
+          format: 1,
+        },
+      ],
+    };
+  }
+  return {
+    path: filePath,
+    hashes_ffm: [
+      {
+        data: digest1,
+        format: 1,
+      },
+      {
+        data: digest2 as string, // digest2 is guaranteed not to be null, due to earlier return
```

Comment:
You can use a non-null assertion in this case: `digest2!`

Although it'd be nicer to fix the types earlier so that its not needed. Perhaps we could try making the return from `getDubHashDigests` more precise? It seems to be either

`[null, null]`, `[string, null]`, `[string, string]`

rather than

`(string | null)[]`

Reply:
Actually, the only possibilities are [string, string] or [string, null]. I've added a type for that.

I tried using non-null assertion, but there seems to be a lint rule against using that (https://github.com/typescript-eslint/typescript-eslint/blob/v3.10.1/packages/eslint-plugin/docs/rules/no-non-null-assertion.md). So unless we disable that rule, I think the casting to string is still necessary.